federation/pkg/federation-controller/sync/controller.go

```diff
@@ -384,6 +391,43 @@ func (s *FederationSyncController) delete(obj pkgruntime.Object, kind string, na
 	return nil
 }
 
+// updateFedSpec updates the spec against the federated object using the adapters UpdateFedSpecHook of need be.
+func (s *FederationSyncController) updateFedSpec(fedObj pkgruntime.Object) reconcileStatus {
+	clusters, err := s.informer.GetReadyClusters()
+	if err != nil {
+		glog.Errorf("Failed to get cluster list: %v", err)
+		return redeliverForClusterReadiness
+	}
+
+	current := make(map[string]pkgruntime.Object)
+	key := federatedtypes.ObjectKey(s.adapter, fedObj)
+	for _, cluster := range clusters {
+		clusterObj, found, err := s.informer.GetTargetStore().GetByKey(cluster.Name, key)
+		if err != nil {
+			glog.Errorf("Failed to get %s %q from target store %q: %v", s.adapter.Kind(), key, cluster.Name, err)
+			return redeliverForFailure
+		}
+		if found {
+			current[cluster.Name] = clusterObj.(pkgruntime.Object)
+		}
+	}
+
+	if len(current) == 0 {
+		return redeliver
+	}
+
+	// The scheduler adapter type implementer implements additional method which is used here.
+	// The implementation should ensure not to mutate any of the passed objects.
+	schedAdapter := s.adapter.(federatedtypes.SchedulerAdapter)
```

Comment:
That somehow did not look clean to me. If you are worried about the method signature being botched up, then it wouldn't just compile. If you are worried about the method implementation, well.. what stops it being wrong even after IsSchedule() saying, yes this adapter has a Schedule() method. Anything else, I am missing?